cheroot/ssl/__init__.py

```diff
@@ -1,16 +1,162 @@
-"""Implementation of the SSL adapter base interface."""
+"""Implementation of the SSL adapter base interface.
+
+.. spelling::
+   dn
+"""
 
 from abc import ABC, abstractmethod
+from contextlib import suppress
+
+
+DN_SEPARATOR = '/'  # Distinguished Name separator
+UTF8_ENCODING = 'utf-8'
+
+
+def _parse_dn_components(components, key_prefix, dn_type):
+    """
+    Parse Distinguished Name components into environ dict.
+
+    Args:
+        components: Iterable of (key, value) tuples
+        key_prefix: 'SSL_CLIENT' or 'SSL_SERVER'
+        dn_type: 'S' for subject or 'I' for issuer
+
+    Returns:
+        dict: ``DN`` and ``CN`` environment variables
+    """
+    env = {}
+    dn_parts = []
+
+    for key, attr_value in components:
+        dn_parts.append(f'{key}={attr_value}')
+        if key in {'CN', 'commonName'}:
+            env[f'{key_prefix}_{dn_type}_DN_CN'] = attr_value
+
+    if dn_parts:
+        dn_string = DN_SEPARATOR.join(dn_parts)
+        env[f'{key_prefix}_{dn_type}_DN'] = f'{DN_SEPARATOR}{dn_string}'
+
+    return env
+
+
+def parse_pyopenssl_cert_to_environ(cert, key_prefix):
+    """Parse a pyOpenSSL X509 certificate into WSGI environ dict."""
+    env = {}
+    if not cert:
+        return env
+
+    # Subject
+    subject = cert.get_subject()
+    if subject:
+        components = [
+            (key.decode(UTF8_ENCODING), attr_value.decode(UTF8_ENCODING))
+            for key, attr_value in subject.get_components()
+        ]
+        env.update(_parse_dn_components(components, key_prefix, 'S'))
+
+    # Issuer
+    issuer = cert.get_issuer()
+    if issuer:
+        components = [
+            (key.decode(UTF8_ENCODING), attr_value.decode(UTF8_ENCODING))
+            for key, attr_value in issuer.get_components()
+        ]
+        env.update(_parse_dn_components(components, key_prefix, 'I'))
+
+    # Version and Serial
+    env[f'{key_prefix}_M_VERSION'] = str(cert.get_version())
+    env[f'{key_prefix}_M_SERIAL'] = str(cert.get_serial_number())
+
+    return env
+
+
+def parse_x509_cert_to_environ(cert, key_prefix):
+    """Parse a cryptography x509 certificate into environ dict."""
+    env = {}
+
+    # Subject
+    with suppress(Exception):
+        subject = cert.subject
+        components = [(attr.oid._name, attr.value) for attr in subject]
+        env.update(_parse_dn_components(components, key_prefix, 'S'))
+
+    # Issuer
+    with suppress(Exception):
+        issuer = cert.issuer
+        components = [(attr.oid._name, attr.value) for attr in issuer]
+        env.update(_parse_dn_components(components, key_prefix, 'I'))
+
+    # Version and Serial
+    with suppress(Exception):
+        env[f'{key_prefix}_M_VERSION'] = str(cert.version.value)
+        env[f'{key_prefix}_M_SERIAL'] = str(cert.serial_number)
+
+    return env
```

Comment:
These helpers look like they could be a separate refactoring PR. Also, I'm not sure if it's a good idea to put adapter-specific things into a shared module.